### ISSUE TITLE
Non-UTF8 output from knitr::knit following a (deliberate) error from dplyr::select

### ISSUE THREAD

**Issue by mdenwood** [yihui/knitr#1805]
---

By filing an issue to this repo, I promise that

- [X] I have fully read the issue guide at https://yihui.name/issue/.
- [X] I have provided the necessary information about my issue.
    - If I'm asking a question, I have already asked it on Stack Overflow or RStudio Community, waited for at least 24 hours, and included a link to my question there.
    - If I'm filing a bug report, I have included a minimal, self-contained, and reproducible example, and have also included `xfun::session_info('knitr')`. I have upgraded all my packages to their latest versions (e.g., R, RStudio, and R packages), and also tried the development version: `remotes::install_github('yihui/knitr')`.
    - If I have posted the same issue elsewhere, I have also mentioned it in this issue.
- [X] I have learned the Github Markdown syntax, and formatted my issue correctly.

I understand that my issue may be closed if I don't fulfill my promises.

---

I want to illustrate to students how to mutate and then select a variable using dplyr verbs:

```{r example1}
library('tidyverse')
df <- tibble(varA = 1:10, varB = 11:20)
df %>% mutate(varC = varA + varB) %>% select(varC)
```

Then I (deliberately) cause an error for demonstration purposes:

```{r example2, error=TRUE}
df %>% select(varC) %>% mutate(varC = varA + varB)
```

When knitting this file from .Rmd to .md, a non-UTF8 character is created around the error message produced by tidyselect (the format of which I think has changed recently to include a little cross symbol). This introduces some rogue characters at the start of the relevant line after subsequent pandoc to HTML and Word, and Tex bails due to the non-UTF8ness. A minimal reproducible example is here (I had to change the extension from .Rmd to .txt to upload):

[knitrbug.txt](https://github.com/yihui/knitr/files/4134703/knitrbug.txt)

Running the same code in a plain R file with output redirected seems to result in valid UTF8 so I guess this is a knitr::knit issue of some kind?  In my case it is easy to workaround by not running the failing example (using eval=FALSE) but this kind of thing can be instructive for teaching purposes.  Hopefully this is fairly self-explanatory but let me know if I can assist with anything.

My sessionInfo (knitr installed from GitHub):

	R version 3.6.2 (2019-12-12)
	Platform: x86_64-apple-darwin15.6.0 (64-bit)
	Running under: macOS High Sierra 10.13.6

	Locale: en_GB.UTF-8 / en_US.UTF-8 / en_GB.UTF-8 / C / en_GB.UTF-8 / en_GB.UTF-8

	Package version:
	  evaluate_0.14   glue_1.3.1      graphics_3.6.2  grDevices_3.6.2
	  highr_0.8       knitr_1.27.2    magrittr_1.5    markdown_1.1
	  methods_3.6.2   mime_0.8        stats_3.6.2     stringi_1.4.5
	  stringr_1.4.0   tools_3.6.2     utils_3.6.2     xfun_0.12
	  yaml_2.2.0

**Comment by shrektan** [comment]
Yes, I can reproduce this with the latest version of tidyverse packages. I believe the cause is the error message contains a special character `'\033[31mx\033[39m'` which can't be handled by the LaTeX engine. I'm thinking about a workaround for it.

**Comment by shrektan** [comment]
@mdenwood The only easy way I can think of is to disable the [crayon](https://github.com/r-lib/crayon) output via `options(crayon.enabled = FALSE)`.

Below is a full example:

``````rmd
---
title: "Bug in knitr::knit when dplyr::select gives a (deliberate) error"
output: pdf_document
---

```{r setup}
knitr::opts_chunk$set(echo = TRUE)
# disable the crayon message
options(crayon.enabled = FALSE)
```

```{r example1}
library('tidyverse')
df <- tibble(varA = 1:10, varB = 11:20)
df %>% mutate(varC = varA + varB) %>% select(varC)
```

```{r example2, error=TRUE}
print(interactive())
df %>% select(varC) %>% mutate(varC = varA + varB)
```
``````

**Comment by shrektan** [comment]
@yihui In my opinion, disabling the `cayron` message automatically seems a good choice to me because I don't think the [ANSI escape code](https://en.wikipedia.org/wiki/ANSI_escape_code) can
be easily handled by LaTeX engine or other formats (HTML will display it as a plain "x", word will display it as confusing texts "[31mx[39m"). In fact, it seems a good choice only when used interactively. 

However, I'm new to this and I may be wrong. Do you agree with my opinion? If so, I would happy to file a PR. Thanks.

**Comment by mdenwood** [comment]
Thanks @shrektan - the suggestion to disable crayon messages works perfectly for me, and I (naively) agree that it seems reasonable to do this by default.

On an semi-related note I always restrict the cli.width as otherwise the tidyverse startup message is wider than the PDF output - I wonder if something like ```options(cli.width = 80)``` as a default setting could be added at the same time?

In any case you have provided a solution to my problem so I am closing this issue.

**Comment by cderv** [comment]
I think there could be something to do with `crayon` at the end. There could be a logic in crayon to know if it is executed in a rmd document and with which output format to deal with the deactivation. 

As pointed by @gaborcsardi at tidydevdat, there is already something in `cli` for unicode characters [detecting the latex output](https://github.com/r-lib/cli/blob/e9f041eaea501301396658078b5dbd38f109865a/R/utf8.R#L18)
I'll open an issue in crayon about that.

**Comment by gaborcsardi** [comment]
Yeah, we can do this in crayon, can you pls open an issue there? THanks!

**Comment by cderv** [comment]
So while trying to work for a reprex to open an issue in crayon, it is not that obvious where this should be fixed. I managed thanks to the help of @lionel- and its rlang expertise to build this reprex 

````
---
title: "crayon in condition message"
output: 
  latex_document: default
---

```{r setup}
# we want to catch error
knitr::opts_chunk$set(error = TRUE)
```

We use special rlang formatting for error.
If the error is signal directly in the chunk, `crayon` gets correctly disabled

```{r}
rlang::abort(rlang:::format_error_bullets(c(x = "dplyr-like error")), .subclass = "foobar")
```

But if we create a lazy condition message like it is done in `tidyselect`, we get the error. 

```{r, error = TRUE}
conditionMessage.foobar <- function(cnd) rlang:::format_error_bullets(c(x = "dplyr-like error"))
rlang::abort(.subclass = "foobar")
```

If we disable crayon, it is working ok. 
```{r}
options(crayon.enabled = FALSE)
rlang::abort(.subclass = "foobar")
```
````

First chunk will render to tex ok. 
```tex
\begin{verbatim}
## Error: x dplyr-like error
\end{verbatim}
```
Second one will have the ansi character causing the issue for pdf rendering by latex
```tex
\begin{verbatim}
## Error: [31mx[39m dplyr-like error
\end{verbatim}
```
Disabling the option is indeed taken into account because the third chunk render ok in tex
```tex

\begin{verbatim}
## Error: x dplyr-like error
\end{verbatim}
```

So it seems this is more complicated that the first crayon issue I had in mind. Now I have this reprex, I can dig into it more.

**Comment by cderv** [comment]
Some updates are waiting for my 🛫 gives me some time. 

The issue is not with latex in fact. The ANSI strings are already there in the rendering to markdown by knitr. I think they shouldn't be there. 

One thing I now understand is that **crayon is correctly disabled by default inside Rmd document rendering**. I am not sure it is completely fully handled but I think this is because of this condition being TRUE [here](https://github.com/r-lib/crayon/blob/f4bc7b852fda22c5f95136e09d26a4aac0ad506a/R/has_color.r#L64)
```r
  ## Are we in a terminal? No?
  if (!isatty(stdout())) { return(FALSE) }
```

Here are the test I made. 

<details><summary>Source</summary>

````r
---
title: "crayon in Rmd"
output: md_document
---

crayon is correclty disabled when rendering Rmarkdown document (but not when running chunks in RStudio IDE) 

```{r}
crayon::red("red text")
withr::with_options(list('crayon.enabled' = TRUE), 
                    crayon::red("red text"))
```

This is indeed ok that rlang error does not contains the ANSI escape sequence

```{r, error = TRUE}
rlang::abort(rlang:::format_error_bullets(c(x = "dplyr-like error")), .subclass = "foobar")
```

But this is not ok that the ANSI escape sequences are not stripped when the condition message is built like in the tidyverse (and so the error above)

```{r, error = TRUE}
conditionMessage.foobar <- function(cnd) rlang:::format_error_bullets(c(x = "dplyr-like error"))
rlang::abort(.subclass = "foobar")
```

However, if we disable crayon explicitly, it is working ok. 

```{r, error = TRUE}
options(crayon.enabled = FALSE)
rlang::abort(.subclass = "foobar")
```

So there is something that is not correctly disabling crayon in this case. 
```` 

</details>

**md results after `rmarkdown::render`**

crayon is correclty disabled when rendering Rmarkdown document (but not
when running chunks in RStudio IDE)

    crayon::red("red text")

    ## [1] "red text"

    withr::with_options(list('crayon.enabled' = TRUE), 
                        crayon::red("red text"))

    ## [1] "\033[31mred text\033[39m"

This is indeed ok that rlang error does not contains the ANSI escape
sequence

    rlang::abort(rlang:::format_error_bullets(c(x = "dplyr-like error")), .subclass = "foobar")

    ## Error: x dplyr-like error

But this is not ok that the ANSI escape sequences are not stripped when
the condition message is built like in the tidyverse (and so the error
above)

    conditionMessage.foobar <- function(cnd) rlang:::format_error_bullets(c(x = "dplyr-like error"))
    rlang::abort(.subclass = "foobar")

    ## Error: [31mx[39m dplyr-like error

However, if we disable crayon explicitly, it is working ok.

    options(crayon.enabled = FALSE)
    rlang::abort(.subclass = "foobar")

    ## Error: x dplyr-like error

So there is something that is not correctly disabling crayon in this
case.

I'll continue to look into it. 💪

**Comment by cderv** [comment]
I have now isolate the issue and I think it is indeed related to crayon., and this is the last minimal reprex:

<details><summary>Rmd source</summary>

````r
---
title: "crayon in Rmd"
output: md_document
---

## With base R

In a Rmd, a condition with a message directly in it that uses `crayon` 
will know it has no color support.

```{r, error = TRUE}
er <- errorCondition(
  message = paste("crayon::has_color() ?:",
                  crayon::has_color(),
                  "\nresult:",
                  crayon::red("A barfoo error")))
stop(er)
```

However, using a specific class for a condition and definining a special method
for `conditionMessage` will not know that color are not disabled.

```{r, error = TRUE}
conditionMessage.barfoo <- function(c) paste("crayon::has_color() ?:",
                                             crayon::has_color(),
                                             "\nresult:",
                                             crayon::red("A barfoo error"))
er <- errorCondition("", class = "barfoo")
stop(er)
```

## With rlang

The above is related to what happened when using rlang and thus tidyverse
:packages: as condition messages are not defined directly inside condition but
evaluated when signalled. (As I understand...)

```{r, error = TRUE}
rlang::abort(
  message = paste(
    "crayon::has_color() ?:",
    crayon::has_color(),
    "\n error message:",
    rlang:::format_error_bullets(c(x = "dplyr-like error"))
    )
)
```

```{r, error = TRUE}
conditionMessage.foobar <- function(cnd) {
  paste(
    "crayon::has_color() ?:",
    crayon::has_color(),
    "\n error message:",
    rlang:::format_error_bullets(c(x = "dplyr-like error"))
  )}
rlang::abort(.subclass = "foobar")
```
````
</details>

With base R
-----------

In a Rmd, a condition with a message directly in it that uses `crayon`
will know it has no color support.

    er <- errorCondition(
      message = paste("crayon::has_color() ?:",
                      crayon::has_color(),
                      "\nresult:",
                      crayon::red("A barfoo error")))
    stop(er)

    ## Error: crayon::has_color() ?: FALSE 
    ## result: A barfoo error

However, using a specific class for a condition and definining a special
method for `conditionMessage` will not know that color are not disabled.

    conditionMessage.barfoo <- function(c) paste("crayon::has_color() ?:",
                                                 crayon::has_color(),
                                                 "\nresult:",
                                                 crayon::red("A barfoo error"))
    er <- errorCondition("", class = "barfoo")
    stop(er)

    ## Error: crayon::has_color() ?: TRUE 
    ## result: [31mA barfoo error[39m

With rlang
----------

The above is related to what happened when using rlang and thus
tidyverse :packages: as condition messages are not defined directly
inside condition but evaluated when signalled. (As I understand…)

    rlang::abort(
      message = paste(
        "crayon::has_color() ?:",
        crayon::has_color(),
        "\n error message:",
        rlang:::format_error_bullets(c(x = "dplyr-like error"))
        )
    )

    ## Error: crayon::has_color() ?: FALSE 
    ##  error message: x dplyr-like error

    conditionMessage.foobar <- function(cnd) {
      paste(
        "crayon::has_color() ?:",
        crayon::has_color(),
        "\n error message:",
        rlang:::format_error_bullets(c(x = "dplyr-like error"))
      )}
    rlang::abort(.subclass = "foobar")

    ## Error: crayon::has_color() ?: TRUE 
    ##  error message: [31mx[39m dplyr-like error

On this basis, I will now open an issue in `crayon` to see how `has_color` should be fixed or modified to handle this case.

**Comment by cderv** [comment]
This has been fixed upstream in r-lib/cli#196 - it is available in dev version of cli and crayon. Will be on CRAN next year 😃
